README.md
# Git-annex special remote for accessing (Remote) Indexed file Archives (RIA)

[![Travis tests status](https://secure.travis-ci.org/datalad/git-annex-ria-remote.png?branch=master)](https://travis-ci.org/datalad/git-annex-ria-remote) [![GitHub release](https://img.shields.io/github/release/datalad/git-annex-ria-remote.svg)](https://GitHub.com/datalad/git-annex-ria-remote/releases/) [![PyPI version fury.io](https://badge.fury.io/py/ria-remote.svg)](https://pypi.python.org/pypi/ria-remote/)

This [git-annex](http://git-annex.branchable.com) [special
remote](http://git-annex.branchable.com/special_remotes) implementation is very
similar to the `directory` special remote type built into git-annex. There are
a few key differences that outline the use cases where one might consider using
this one instead:

- (Optional) read-access to (compressed) 7z archives

  (Parts of) the keys stored in the remote can live in a
  [7z](https://www.7-zip.org) archive. These archives are indexed and support
  relatively fast random read access. This feature can be instrumental on
  HPC storage systems where strong quotas on filesystem inodes might be imposed
  on users. The entire key store of the remote can be put into an archive, re-using
  the exact same directory structure, and remains fully accessible while only
  using a handful of inodes, regardless of file number and size.

- (SSH-based remote) access to a configurable directory

  An SSH host name can be provided and all interaction with the remote will be
  performed via SSH. Moving from local to remote operations, or switching target
  paths can be done via a change to the configuration (even without having to touch
  a repository at all). This makes it easier to accommodate infrastructural changes,
  especially when dealing with large numbers of repositories.

- Multi-repository directory structure

  While each repository has its own associated key store directory tree, the
  key store directories of multiple repositories can be organized into a homogeneous
  archive directory structure. For [DataLad](http://datalad.org) datasets, their
  ID is used to define the location of a key store in an archive. For any other
  repository the annex remote UUID is taken. This feature further aids the handling
  of large numbers of repositories in a backup or data store use case, because
  locations are derived from repository properties rather than having to re-configure
  them explicitly.


## Installation

Before you install this package, please make sure that you [install a recent
version of git-annex](https://git-annex.branchable.com/install).  This special
remote requires at minimum git-annex version 6.20160511. Afterwards,
install the latest version of `ria-remote` from
[PyPi](https://pypi.org/project/ria-remote):

    # install from PyPi
    pip install ria-remote


## Use

A `ria` special remote is set up like any other "external"-type remote via the
git-annex `initremote` command. There is a single additional required setting
in contrast to the standard ones: `base-path` which determines the base
directory where the special remote places its keys:

    git annex initremote myremote \
        type=external encryption=none \
        externaltype=ria base-path=/tmp/basepath/here

Alternatively, the `base-path` can also be provided via a Git configuration
variable by setting `annex.ria-remote.<remote>.base-path` (in this example
`annex.ria-remote.myremote.base-path`).

The remote is now ready for use. Any directories will be created on demand.
The key store for a repository will be located underneath the given base path,
in a structure like this:

    /tmp/basepath/here
    └── 2e5
        └── 24934-a09e-11e9-8503-f0d5bf7b5561
            └── annex
                └── objects
                    └── ff4
                        └── c57
                            └── MD5E-s4--ba1f2511fc30423bdbb183fe33f3dd0f
                                └── MD5E-s4--ba1f2511fc30423bdbb183fe33f3dd0f

where the first two levels represent a tree structure that can host key stores
for any number of repositories, and the remaining level are identical to
the organization of a bare Git repository with the annex object tree following
the layout of a `directory`-type git-annex special remote.
The directory names for the two top-most levels are build from the git-annex
UUID for the special remote, or a DataLad dataset UUID, if available.

The special remote also supports SSH-based operation. To enable it, an
additional host name argument has to be given:

    git annex initremote myremote \
        type=external encryption=none \
        externaltype=ria base-path=/tmp/basepath/here \
        ssh-host=ria.example.com

This configuration will make the special remote use `/tmp/basepath/here` on
`ria.example.com`. Any SSH-access customizations (user name, ports, etc.) have
to be implemented via the standard SSH configuration mechanism, for example, by
placing a snippet like this in `$HOME/.ssh/config`:

    Host ria.example.com
      User mike
      Port 2222
      PreferredAuthentications publickey

There are additional configuration settings available:

- By default ria-remote will check the remote end's layout version by reading a
  `ria-layout-version` file at the top-level (`base-path`) as well as at the
  individual dataset directories. If the layout version isn't known it will set
  its mode to "read-only" and will reject to write anything to that store in
  order to not accidentally mix up different layouts. This behavior can be
  overwritten by setting `annex.ria-remote.<name>.force-write` to `true`.

- The remote end can indicate that any occurring exception shall be written to
  a log file. This can be helpful for debugging problems in a multi-user
  central storage scenario. For this to be used, the version recorded in the
  `ria-layout-version` file can be appended by `|l` ('l' for logging). If done at
  the top of the dataset tree, this will applied to all datasets.  However, since
  such a log can potentially leak private information, this configuration can be
  ignored by any client by setting
  `annex.ria-remote.<name>.ignore-remote-config`. There is no independent
  server-side processing, and all actions are performed by the client-side
  special remote instance.

## Support

All bugs, concerns and enhancement requests for this software can be submitted here:
https://github.com/datalad/git-annex-ria-remote/issues


CHANGELOG
0.7 (Dec 18, 2019) -- Easy Access

- add `ria-install` command to install a dataset directly from a given store
- discard the dirhashlower approach re annex/objects


0.6.1 (Nov 13, 2019) -- Grrrr

- Disabled package metadata to get upload to PyPi done.

0.6 (Nov 12, 2019) -- Only for export

- New DataLab subcommand 'ria-export-archive' for exporting a local annex
  object store into a 7z archive for use with the RIA special remote.
- Ongoing data transfers are now using a remote-specific temporary directory
  to prevent collisions.

0.5 (Sep 27, 2019) -- Speed!

- Communication with an SSH host is now done via a persistent remote shell
  session. This massively reduces the command latency on systems that are
  slow in providing a new session.
- Further minimization of the number of commands that are executed on
  remote systems.

0.4 (Aug 20, 2019) -- Tests!

- Package now called 'ria-remote' on PyPi
- Make data store layout compatible with a co-located bare repository,
  by moving the object tree under 'annex/objects'
- Add layout version files to the global data store, and to each
  dataset directory. For any newly created store they will be added
  automatically. For existing stores the special remote will refuse
  to write into unversioned locations, unless forced
  (config variable `force-write`)
- Fix issue with inappropriate exception type that `annexremote` did not
  convert into a proper message for git-annex
- Reduce number of required SSH invokations for 'checkpresent'
- Significantly expanded test coverage for various use cases

0.3 (Jul 09, 2019) -- ready?

- All functionality implemented
- Tests for major use cases, and git-annex compatibility
- Travis CI setup
- Some documentation

0.2 (May 15, 2019) -- make it run

- All desired functionality for a usable draft

0.1 (May 14, 2019) -- don't touch this

- Initial draft implementation


ria_remote/resources/procedures/ria_post_install.py
"""

"""

import logging
import sys
import os
import posixpath
from pathlib import (
    PosixPath,
    Path
)
from datalad.distribution.dataset import require_dataset
from datalad.utils import rmtree
from ria_remote import RIARemote


lgr = logging.getLogger('datalad.procedure.ria_post_install')


def proc_ria_remote(ds, rm, origin, special_remotes, ephemeral, reckless):
    special_remote_info = special_remotes.get(rm['annex-uuid'], None)
    # base-path must be somewhere, otherwise it could not be active
    base_path = ds.config.get(
        'annex.ria-remote.{}.base-path'.format(rm['name']), None)
    sshhost = ds.config.get(
        'annex.ria-remote.{}.ssh-host'.format(rm['name']), None)
    if not base_path:
        base_path = special_remote_info.get('base-path', None)
    if not base_path:
        lgr.error(
            'ria_post_install logic error: active RIA special '
            'remote without base-path setting')
        return
    dspath_in_ria, _, _ = RIARemote.get_layout_locations(Path(base_path), ds.id)
    if not (origin_remote['url'].endswith(str(PosixPath(dspath_in_ria))) or
            origin_remote['url'] == str(dspath_in_ria)):
        # we have no business here. neither a local, nor a remote clone
        # from this RIA store
        return

    ds.siblings(
        'configure',
        name='origin',
        publish_depends=rm['name'],
        result_filter=None,
        result_renderer='disabled')

    ds.config.set(
        'datalad.clone.proc-post', 'ria_post_install {} {}'.format(
            'ephemeral' if ephemeral else '',
            'reckless' if reckless else '',
        ),
        where='local')

    ds.config.set(
        'datalad.get.subdataset-source-candidate-{}'.format(rm['name']),
        '{host}{path}'.format(
            host=sshhost + ':' if sshhost else '',
            # TODO think about making this work in windows too
            path=PosixPath(base_path) / '{id:.3}/{id[3]}{id[4]}{id[5]}{id[6]}{id[7]}{id[8]}{id[9]}{id[10]}{id[11]}{id[12]}{id[13]}{id[14]}{id[15]}{id[16]}{id[17]}{id[18]}{id[19]}{id[20]}{id[21]}{id[22]}{id[23]}{id[24]}{id[25]}{id[26]}{id[27]}{id[28]}{id[29]}{id[30]}{id[31]}{id[32]}{id[33]}{id[34]}{id[35]}',
        ),
        where='local')

    if ephemeral:
        # with ephemeral we declare 'here' as 'dead' right away, whenever
        # we symlink origins annex. Because we want annex to copy to
        # the ria remote to get availability info correct for an eventual
        # git-push into the store
        # this will cause stuff like this for a locally present annexed file:
        # % git annex whereis d1
        # whereis d1 (0 copies) failed
        # BUT this works:
        # % git annex find . --not --in here
        # % git annex find . --in here
        # d1

        # we don't want annex copy-to origin
        ds.config.set(
            'remote.origin.annex-ignore', 'true',
            where='local')

        ds.repo._run_annex_command('dead', annex_options=['here'])

        if reckless and origin_remote['url'] == str(dspath_in_ria):
            # cloned from a RIA store at a local path, symlink the annex
            # to avoid needless copies in an emphemeral clone
            annex_dir = ds.repo.dot_git / 'annex'
            # TODO make sure that we do not delete any unique data
            rmtree(str(annex_dir)) \
                if not annex_dir.is_symlink() else annex_dir.unlink()
            annex_dir.symlink_to(
                os.path.join(origin_remote['url'], 'annex'),
                target_is_directory=True)


ds = require_dataset(
    sys.argv[1],
    check_installed=True,
    purpose='Post-installation setup for datasets installed from a RIA store')

ephemeral = len(sys.argv) > 2 and 'ephemeral' in sys.argv[2:]
reckless = len(sys.argv) > 2 and 'reckless' in sys.argv[2:]

ria_remotes = [
    s for s in ds.siblings('query', result_renderer='disabled')
    if s.get('annex-externaltype', None) == 'ria'
]

if not ria_remotes:
    lg = lgr.warning if any(
        # TODO also check for base-path and compare against clone URL
        s.get('externaltype', None) == 'ria'
        for s in ds.repo.get_special_remotes().values()) else lgr.debug
    lg('No active RIA remote found')
    # we don't want to fail, this could all be valid
    sys.exit(0)

origin_remote = ds.siblings(
    'query', name='origin', return_type='item-or-list',
    result_renderer='disabled')

special_remotes = ds.repo.get_special_remotes()

for rm in ria_remotes:
    proc_ria_remote(
        ds, rm, origin_remote, special_remotes, ephemeral, reckless)
